### PR TITLE
[DEV-994] fixed loose validator checks

### PR DESCRIPTION
**Pull request by Krysh90**

### CODE REVIEW

File: shared/transaction/validator.ts [hunk 1]
```diff
@@ -32,34 +32,45 @@ export class Validator {
...
+  private static voutHas(vout: Vout, script: Script): boolean {
+    if (vout.script.stack.length !== script.stack.length) return false;
+    return vout.script.stack.every(
+      (code, index) => code.asBuffer().toString() === script.stack[index].asBuffer().toString(),
+    );
+  }
+
+  private static voutHasWithAmount(vout: Vout[], value: BigNumber, opCodes: OPCode[]): boolean {
     return vout.filter((v) => v.value.isEqualTo(value) && this.includesAll(v.script.stack, opCodes)).length === 1;
   }
```
Comment (davidleomay) [inline shared/transaction/validator.ts]: The naming of these two methods makes it hard to understand what is going on 🤔

Comment (davidleomay) [inline shared/transaction/validator.ts]: What do we actually verify when comparing the stack?

File: shared/transaction/validator.ts [hunk 2]
```diff
@@ -32,34 +32,45 @@ export class Validator {
...
-  private static voutHas(vout: Vout[], value: BigNumber, opCodes: OPCode[]): boolean {
+  private static voutHas(vout: Vout, script: Script): boolean {
+    if (vout.script.stack.length !== script.stack.length) return false;
+    return vout.script.stack.every(
+      (code, index) => code.asBuffer().toString() === script.stack[index].asBuffer().toString(),
+    );
+  }
+
+  private static voutHasWithAmount(vout: Vout[], value: BigNumber, opCodes: OPCode[]): boolean {
     return vout.filter((v) => v.value.isEqualTo(value) && this.includesAll(v.script.stack, opCodes)).length === 1;
```
Comment (davidleomay) [inline shared/transaction/validator.ts]: Does this check work with objects (OPCodes)?

File: shared/transaction/validator.ts [hunk 3]
```diff
@@ -32,34 +32,45 @@ export class Validator {
...
         OP_CODES.OP_DEFI_TX_RESIGN_MASTER_NODE(undefined as unknown as ResignMasternode),
       ])
     );
   }
 
   private static sendFromLiq(tx: CTransactionSegWit, script: Script): boolean {
-    return tx.vout.length === 2 && this.voutContains(tx.vout, [[OP_CODES.OP_RETURN], script.stack]);
+    return (
+      tx.vout.length === 2 &&
+      !tx.vout[0].script.stack.map((code) => `${code.type}`).includes('OP_DEFI_TX') &&
```
Comment (davidleomay) [inline shared/transaction/validator.ts:54]: Is this a blockchain rule that the DeFi TX is always the first vout?

Comment (Krysh90) [reply shared/transaction/validator.ts:54]: yes. that is why there can only be one DEFI_TX inside one tx. As they need to be at index 0

File: shared/transaction/validator.ts (hunk 1 above)
Comment (Krysh90) [reply shared/transaction/validator.ts]: will rename them

Comment (Krysh90) [reply shared/transaction/validator.ts]: inside the script.stack the public key hash of the address is encoded inside. By checking the stack, we can verify that the vout destination is a specific address

File: shared/transaction/validator.ts (hunk 2 above)
Comment (Krysh90) [reply shared/transaction/validator.ts]: got reworked

**Review by davidleomay — APPROVED** [review]

**Pull request merged**